crates/wysiwyg/src/tests/test_deleting.rs

```diff
@@ -222,3 +222,37 @@ fn deleting_a_newline_deletes_it() {
     model.delete();
     assert_eq!(tx(&model), "abc|ef");
 }
+
+#[test]
+fn test_backspace_emoji() {
+    let mut model = cm("😄|😅");
+    model.backspace();
+    assert_eq!(tx(&model), "|😅");
+}
+
+#[test]
+fn test_backspace_complex_emoji() {
+    let mut model = cm("Test😮‍💨|😅");
+    model.backspace();
+    assert_eq!(tx(&model), "Test|😅");
```

Comment:
Are we sure we want this behaviour?

In the github editor, if I paste in 😮‍💨 and then backspace once I get 😮. So it only deleted one codepoint.

I think the crash came because we tried to delete a partial codepoint, not a partial grapheme (but I'm not sure)

Reply:
If we want to delete individual codepoints (which I think we do) we can use `String.chars` instead of `graphemes`.

If we want to delete graphemes, I think we should use https://crates.io/crates/unicode-segmentation which has a `graphemes` method. The code we have here only supports zero-width-joiners, but there are lots of "combining" characters that don't need a joiner - they just combine automatically (e.g. Combining Acute Accent (U+0301)). I assume unicode-segmentation handles those.

Reply:
> Are we sure we want this behaviour?

On Android at least, that's the expected behaviour. If I paste `😮‍💨` here (GH editor) and I backspace the whole emoji gets deleted, so maybe it's a per OS implementation?

> I think the crash came because we tried to delete a partial codepoint

Yes, that seems to be the case.

> If we want to delete graphemes, I think we should use https://crates.io/crates/unicode-segmentation which has a graphemes method.

`unicode-segmentation` only works for UTF-8 strings. I tried using it at first but as soon as I passed a UTF-16 string everything crashed. That's why I implemented this other solution, trying to use it as an inspiration.

> The code we have here only supports zero-width-joiners, but there are lots of "combining" characters that don't need a joiner

To be honest, I haven't actually tried those. I'll do and come back with my results.

Reply:
As you thought, the code only handles emoji properly, for those combining characters, each character is treated as a different cluster.

I took a look at the `UnicodeSegmentation` trait, trying to figure out if we could implement it for `Utf16Str` and `Utf32Str`, but it turns out it's hardcoded to return a `Graphemes` struct that contains a `str`, so the crate was only thought to work with Rust strings.

I could *try* to mimick the implementation of the crate or even reuse some of its utils to improve the logic used here, but it would take quite some time and I think there are more pressing issues at the moment. Would we be happy using this to fix the emoji crashes with a TODO to properly support graphemes in the future?

Reply:
I think I'd prefer to just delete individual code points, rather than provide something that looks like it works, but then doesn't in some other circumstances. What do you think?

Reply:
If we're really committed to graphemes, I'd want to use UnicodeSegmentation, and convert back and forth to/from UTF-8.

Reply:
> I think I'd prefer to just delete individual code points, rather than provide something that looks like it works, but then doesn't in some other circumstances. What do you think?

I think that's probably the best compromise at the moment.

> If we're really committed to graphemes, I'd want to use UnicodeSegmentation, and convert back and forth to/from UTF-8.

I guess this is possible too, but we'd have to map the indexes between the encodings very carefully. I can give this a try and if I see that it takes too much time, go back to just handling codepoints.

Reply:
I would suggest just code points for now. If we want to do graphemes we probably need to discuss with product. We might want to match platform behaviour, or we might want to be consistent.